### PR TITLE
ISPN-10310 State provider should not block while sending state

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.jboss.org/browse/ISPN-10310

Second part of the series, handling topology updates without blocking is still pending.
I have made some changes to XSiteStateProviderImpl to make it more like StateProviderImpl, but I did not make it non-blocking.

### CODE REVIEW

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManager.java [hunk 1]
```diff
@@ -193,7 +193,9 @@
 
    /**
     * Returns the store one configured with fetch persistent state, or null if none exist.
+    * @deprecated Since 10.0, no longer used
     */
+   @Deprecated
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/manager/PersistenceManager.java]: Yay! :)

File: core/src/main/java/org/infinispan/configuration/cache/Configurations.java [hunk 2]
```diff
@@ -95,6 +95,10 @@ public static boolean isClustered(GlobalConfiguration globalConfiguration) {
       return globalConfiguration.transport().transport() != null;
    }
 
+   public static boolean isStateTransferStore(StoreConfiguration storeConfiguration) {
+      return !storeConfiguration.shared() && storeConfiguration.fetchPersistentState();
```
Comment (wburns) [inline core/src/main/java/org/infinispan/configuration/cache/Configurations.java]: I wonder if instead we should have a validation check of cache store config to not allow `fetchPersistentState` and `shared` together as they don't make any sense and throw an exception. Then this just becomes a simple check of `fetchPersistentState`.

File: core/src/main/java/org/infinispan/reactive/RxJavaInterop.java [hunk 3]
```diff
@@ -124,4 +134,35 @@ private RxJavaInterop() { }
    };
 
    private static final Function<Map.Entry<Object, Object>, Object> entryToKeyFunction = Map.Entry::getKey;
+
+   private static final java.util.function.Function<CompletionStage<?>, Completable>
+      completionStageCompletableFunction = CompletionStageCompletableAdapter::new;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/reactive/RxJavaInterop.java]: I think using the following might be a bit more clear.

```
  = stage -> {
        CompletableSubject cs = CompletableSubject.create();

        future.whenComplete((v, e) -> {
            if (e != null) {
                cs.onError(e);
            } else {
                cs.onComplete();
            }
        });

        return cs;
     }
```

File: core/src/main/java/org/infinispan/scattered/impl/ScatteredStateProviderImpl.java [hunk 4]
```diff
@@ -75,34 +80,40 @@ public void start() {
...
+         OutboundTransferTask outboundTransferTask =
+            new OutboundTransferTask(nextMember, oldSegments, cacheTopology.getCurrentCH().getNumSegments(), chunkSize,
+                                     cacheTopology.getTopologyId(), keyPartitioner,
+                                     chunks -> invalidateChunks(chunks, otherMembers, outboundInvalidations,
+                                                                invalidationFuture, cacheTopology),
+                                     OutboundTransferTask::defaultMapEntryFromDataContainer,
+                                     OutboundTransferTask::defaultMapEntryFromStore,
+                                     dataContainer, persistenceManager, rpcManager, commandsFactory, entryFactory,
+                                     timeout, cacheName, true, true);
+         outboundTransferTask.execute().whenComplete((ignored, throwable) -> {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/scattered/impl/ScatteredStateProviderImpl.java]: We probably want to log something if `throwable` is not null.

Comment (wburns) [reply core/src/main/java/org/infinispan/scattered/impl/ScatteredStateProviderImpl.java]: I guess it isn't needed as we swallow all the exceptions further in, I have to admit it feels a little wrong though.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java [hunk 5]
```diff
@@ -162,119 +148,155 @@ public int getTopologyId() {
...
+         if (cancelled) {
+            // In case cancel() was called before we set the subscription field
+            subscription.dispose();
+         }
+         return taskFuture;
+      } catch (Throwable t) {
+         logError(t);
+         taskFuture.complete(null);
+      }
+      return null;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]:
```suggestion
      return taskFuture;
```

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java [hunk 6]
```diff
@@ -162,119 +148,155 @@ public int getTopologyId() {
...
-                        if (segments.contains(segmentId)) {
-                           try {
-                              InternalCacheEntry entry = mapEntryFromStore.apply(me, entryFactory);
-                              if (entry != null) {
-                                 sendEntry(entry, segmentId);
-                              }
-                           } catch (CacheException e) {
-                              log.failedLoadingValueFromCacheStore(me.getKey(), e);
+   private Flowable<InternalCacheEntry<Object, Object>> publishDataContainerEntries() {
+      return Flowable.fromIterable(() -> dataContainer.iterator(segments))
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: One thing to note is that `InternalDataContainer` has a `publisher(int)` method currently. But maybe we should ignore that fact for now, as I hope to utilize that in https://issues.jboss.org/browse/ISPN-10706

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java [hunk 7]
```diff
@@ -162,119 +148,155 @@ public int getTopologyId() {
...
+                     .flatMapMaybe(ice -> {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: I wonder if is probably worth it to be smart and have a static variable that holds `OutboundTransferTask::defaultMapEntryFromDataContainer` and detect if the passed in `mapEntryFromDataContainer` is the same and just do a `filter` and `map` operation instead of a `flatMapMaybe`. We can avoid N allocations of Maybe then.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java [hunk 8]
```diff
@@ -162,119 +148,155 @@ public int getTopologyId() {
...
-      }
-      if (trace) {
-         log.tracef("Completed outbound transfer to node %s, segments %s", destination, segments);
-      }
+   private Flowable<InternalCacheEntry<Object, Object>> publishStoreEntries() {
+      Publisher<MarshallableEntry<Object, Object>> loaderPublisher =
+         persistenceManager.publishEntries(segments, k -> !dataContainer.containsKey(k), true, true,
+                                           Configurations::isStateTransferStore);
+      return Flowable.fromPublisher(loaderPublisher)
+                     .flatMapMaybe(me -> {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: Same here regarding not having to allocate Maybe instances.

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java [hunk 9]
```diff
@@ -206,47 +206,42 @@ public void run() {
...
+                          if (trace) {
+                             log.tracef("Added key '%s' to current chunk", ice.getKey());
+                          }
+                          return Maybe.just(XSiteState.fromDataContainer(ice));
+                       })
+                       .buffer(chunkSize)
+                       .takeUntil(batch -> canceled)
+                       .blockingForEach(batch -> {
+                          try {
+                             // TODO Make non-blocking and use concatMapCompletable like OutboundTransferTask
```
Comment (wburns) [inline core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: Is this going to be covered in a different PR or were you planning on in this one?

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java [hunk 10]
```diff
@@ -206,47 +206,42 @@ public void run() {
...
-               }
-               if (shouldSendKey(ice.getKey())) {
-                  if (trace) {
-                     log.tracef("Added key '%s' to current chunk", ice.getKey());
-                  }
-                  chunk.add(XSiteState.fromDataContainer(ice));
-               }
+            try {
+               Flowable.fromIterable(dataContainer)
+                       .flatMapMaybe(ice -> {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: I would prefer seeing `filter` and `map` here to avoid all the Maybe allocations.

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java [hunk 11]
```diff
@@ -260,12 +255,13 @@ public void run() {
...
-                        .buffer(chunkSize <= 0 ? Integer.MAX_VALUE : chunkSize)
-                        // We want the CacheException to be thrown to the catch block
-                        .blockingForEach(l -> {
+                          .map(XSiteState::fromCacheLoader)
+                          .takeUntil(l -> canceled)
+                          .buffer(chunkSize)
+                          // We want the CacheException to be thrown to the catch block
+                          .blockingForEach(l -> {
                            try {
+                              // TODO Make non-blocking and use concatMapCompletable like OutboundTransferTask
```
Comment (wburns) [inline core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: Is this going to be covered in a different PR or were you planning on in this one?

File: core/src/main/resources/schema/infinispan-config-10.0.xsd [hunk 12]
```diff
@@ -1802,10 +1802,7 @@
         <xs:complexType>
           <xs:attribute name="chunk-size" type="xs:int" default="512">
             <xs:annotation>
-              <xs:documentation>
-                If &gt; 0, the state will be transferred in batches of {@code chunkSize} cache entries.
-                If &lt;= 0, the state will be transferred in all at once. Not recommended. Defaults to 512.
-              </xs:documentation>
+               <xs:documentation>The number of cache entries to batch in each transfer.</xs:documentation>
```
Comment (wburns) [inline core/src/main/resources/schema/infinispan-config-10.0.xsd]: We should document in the upgrading guide that `chunk-size` has to be greater than 0 now.

File: core/src/main/java/org/infinispan/configuration/cache/Configurations.java (hunk 2 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/configuration/cache/Configurations.java]: Sounds good. I didn't think that would work because `fetchPersistentState` is enabled by default and we want to allow users enabling `shared` without also disabling `fetchPersistentState` explicitly, but I was wrong, `fetchPersistentState` is disabled by default.

File: core/src/main/java/org/infinispan/reactive/RxJavaInterop.java (hunk 3 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/reactive/RxJavaInterop.java]: I didn't know about `CompletableSubject`, thanks.

BTW I made it a lambda to match the others, but I don't really need a function object, and I noticed that most callers of the other functions in `RxJavaInterop` also call `apply()` immediately.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java (hunk 6 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: To clarify, we can worry about this later and I am fine with how it is atm.

File: core/src/main/java/org/infinispan/reactive/RxJavaInterop.java (hunk 3 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/reactive/RxJavaInterop.java]: An example is https://github.com/infinispan/infinispan/blob/0d56cc6f8c1556b71df6c1add0d5bf8ab4b68a38/core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java#L382 So if you don't expect to do that, I wouldn't worry about adding the Function.

File: core/src/main/java/org/infinispan/scattered/impl/ScatteredStateProviderImpl.java (hunk 4 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/scattered/impl/ScatteredStateProviderImpl.java]: I was debating whether I should do `thenAccept` instead when I wrote the code... I think I'll move the logging here instead.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java (hunk 6 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: Yeah, I'll ignore it for now :)

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java (hunk 7 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: I'm not even sure we need the filtering, I can't tell if a scattered cache entry is allowed to not have a version, but IMO the allocation is not big enough a deal to add a special case.

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java (hunk 9 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: No plans to tackle xsite in this PR.

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java (hunk 11 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: Different Pr

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManager.java (hunk 1 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/manager/PersistenceManager.java]: Actually looking again, it looks like we can just remove this method as the package is private.

File: core/src/main/resources/schema/infinispan-config-10.0.xsd (hunk 12 above)
Comment (danberindei) [reply core/src/main/resources/schema/infinispan-config-10.0.xsd]: Done

File: core/src/main/java/org/infinispan/reactive/RxJavaInterop.java (hunk 3 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/reactive/RxJavaInterop.java]: I can't make it a regular function, I'd have to move it to another class or give it a new name and I can't think of a new one to explain that it's not like the others.

I knew about `to()`, and I know explicitly caching the lambdas means there's only one callsite/anonymous class/classloader, but I still don't like the explicit `apply()` :)

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java (hunk 10 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: Done

File: core/src/main/java/org/infinispan/scattered/impl/ScatteredStateConsumerImpl.java [hunk 13]
```diff
@@ -280,37 +281,35 @@ protected void onTaskCompletion(InboundTransferTask inboundTransfer) {
...
-                     });
-            } catch (CacheException e) {
-               log.failedLoadingKeysFromCacheStore(e);
-            }
+         Publisher<MarshallableEntry<Object, Object>> persistencePublisher =
+            persistenceManager.publishEntries(completedSegments, k -> !dataContainer.containsKey(k), true, true,
+                                              StoreConfiguration::fetchPersistentState);
+         try {
+            Flowable.fromPublisher(persistencePublisher)
+                    .blockingForEach(me -> {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/scattered/impl/ScatteredStateConsumerImpl.java]: Actually now that I think about this, this is actually blocking. Not because of the invalidation - but because the persistence manager underneath can take time between publishing entries. We need to have a non blocking call here and wait for its completion.

A `blockingForEach` is essentially just a blocking iterator that invokes `forEachRemaining`. So if the entries are not returned in the same thread, we will end up blocking.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java [hunk 14]
```diff
@@ -162,119 +148,138 @@ public int getTopologyId() {
...
 
-   private void sendEntry(InternalCacheEntry ice, int segmentId) {
-      // send if we have a full chunk
-      if (accumulatedEntries >= chunkSize) {
-         sendEntries(false);
-         accumulatedEntries = 0;
-      }
+   private Flowable<InternalCacheEntry<Object, Object>> publishDataContainerEntries() {
+      return Flowable.fromIterable(() -> dataContainer.iterator(segments))
+                     .flatMapMaybe(ice -> {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: Are you sure the Maybe allocations are not worth removing? That would require N Maybe allocations as well as invoking a `flatMap` method, which has its own complications and additional allocations.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java [hunk 15]
```diff
@@ -162,119 +148,138 @@ public int getTopologyId() {
...
+                       })
+                    .subscribe(() -> {
+                       // Send the remaining entries and mark all the segments as finished
+                       List<InternalCacheEntry<Object, Object>> previousBatch = batchRef.get();
+                       sendEntries(previousBatch, true)
+                          .whenComplete((ignored, throwable) -> {
+                             taskFuture.complete(null);
+                          });
+                    }, throwable -> {
+                       taskFuture.complete(null);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: We should probably invoke `completeExceptionally` passing the throwable.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java [hunk 16]
```diff
@@ -162,119 +148,138 @@ public int getTopologyId() {
...
+                             return Completable.complete();
+
+                          return completionStageToCompletable().apply(sendEntries(previousBatch, false));
+                       })
+                    .subscribe(() -> {
+                       // Send the remaining entries and mark all the segments as finished
+                       List<InternalCacheEntry<Object, Object>> previousBatch = batchRef.get();
+                       sendEntries(previousBatch, true)
+                          .whenComplete((ignored, throwable) -> {
+                             taskFuture.complete(null);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: To be safe we should probably invoke `completeExceptionally` here if `throwable` was not null.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java [hunk 17]
```diff
@@ -162,119 +148,138 @@ public int getTopologyId() {
...
-                     });
-            } catch (CacheException e) {
-               log.failedLoadingKeysFromCacheStore(e);
-            }
+         AtomicReference<List<InternalCacheEntry<Object, Object>>> batchRef =
+            new AtomicReference<>(Collections.emptyList());
+         subscription =
+            Flowable.concat(publishDataContainerEntries(), publishStoreEntries())
+                    .buffer(chunkSize)
+                    .concatMapCompletable(
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: I took a peek and this method does a prefetch of 2 elements. I believe this is fine, but I just wanted to point that out.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java [hunk 18]
```diff
@@ -162,119 +148,138 @@ public int getTopologyId() {
...
-               log.tracef("Ignoring error in already cancelled transfer to node %s, segments %s", destination,
-                          segments);
-            }
-         } else {
-            log.failedOutBoundTransferExecution(t);
-         }
-      }
-      if (trace) {
-         log.tracef("Completed outbound transfer to node %s, segments %s", destination, segments);
+         taskFuture.complete(null);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: Why not `completeExceptionally` here?

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java [hunk 19]
```diff
@@ -260,12 +252,13 @@ public void run() {
...
-                        .map(XSiteState::fromCacheLoader)
-                        .takeUntil(l -> canceled)
-                        .buffer(chunkSize <= 0 ? Integer.MAX_VALUE : chunkSize)
-                        // We want the CacheException to be thrown to the catch block
-                        .blockingForEach(l -> {
+                          .map(XSiteState::fromCacheLoader)
+                          .takeUntil(l -> canceled)
+                          .buffer(chunkSize)
+                          // We want the CacheException to be thrown to the catch block
+                          .blockingForEach(l -> {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: Just an FYI, but this is actually blocking irrespective of the Consumer, due to the persistence manager not returning entries in the invoking thread. Just FYI, we may want to comment that so it isn't forgotten for future.

File: core/src/main/java/org/infinispan/scattered/impl/ScatteredStateConsumerImpl.java (hunk 13 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/scattered/impl/ScatteredStateConsumerImpl.java]: I was writing about the `invalidate()` call, because I remember reading this code and thinking we really shouldn't send the entries/invalidations in a `blockingForEach()` call, but I see now I made it more confusing :)

I only meant to remove the usage of `persistenceManager.getStateTransferProvider()`, so I'll remove the comment. I'm going to have another look at the scattered state consumer in another PR, because I also want to limit the number of in-flight invalidation/backup commands.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java (hunk 14 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: I was thinking 1 allocation per entry isn't that bad, especially since it's collected almost immediately. Reading `FlowableFlatMapMaybe` though, looks like you're right and it creates plenty of other objects.

Anyway, I went and changed things so that `flatMapMaybe()` isn't needed anyway :)

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java (hunk 16 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: `sendEntries()` never throws, but ok

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java (hunk 17 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: Yeah, I remember reading how it requests new elements when the prefetch queue is half-depleted, so I figured 2 is about the same as 1. But now that you mentioned it I thought more about it and changed it to 1 to limit memory usage.

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java (hunk 19 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: Yeah, I didn't intend to change anything about xsite in this PR.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java (hunk 16 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: I always like to think that as well, until it does :) And then we lost the exception, making it very difficult to figure out what is going on.

**Comment by wburns** [comment]
Adding Changes Required label until you are able to figure out the blocking in regards to the invalidation commands.

**Comment by danberindei** [comment]
I didn't mean to do anything about the invalidation commands here, sorry for the confusion.
I pushed an update to fix the conflicts.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java (hunk 18 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: Was this missed?

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java [hunk 20]
```diff
@@ -206,91 +207,83 @@ public void run() {
...
+                       .buffer(chunkSize)
+                       .takeUntil(batch -> canceled)
+                       .subscribeOn(Schedulers.from(executorService))
+                       .blockingForEach(batch -> {
+                          try {
+                             // TODO Make non-blocking and use concatMapCompletable like OutboundTransferTask
+                             sendFromSharedBuffer(xSiteBackup, batch, this);
+                          } catch (Throwable t) {
+                             // This will terminate the flowable early
+                             throw new CacheException(t);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: Actually this is a problem now that we are using a Scheduler. It means the exception will be propagated up to the Thread uncaughtException handler. We should just log the error directly in this method instead now.

Comment (wburns) [reply core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: you could just log the error and set `cancelled` to true.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java (hunk 18 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: Yep

File: core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java (hunk 20 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: Right... I'm struggling to remember why I added a scheduler now ;)

Comment (wburns) [reply core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java]: ;)

**Comment by danberindei** [comment]
I squashed the commits and I made two more changes: 
* I added a try..catch in `sendEntries` to catch `IllegalLifecycleException` when the cache manager is stopping
* Instead of disposing the subscription I'm using `takeUntil(batch -> cancelled)`. Initially I wanted to keep only the subscription and remove the `cancelled` field, but maintaining the subscription in a thread-safe way is kind of annoying so I went the other way around.

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java [hunk 21]
```diff
@@ -162,133 +122,139 @@ public int getTopologyId() {
...
+                   // Send the previous batch, not the current one
+                   // This allows us to mark all the segments as finished in the same RPC with the
+                   // last batch
+                   List<InternalCacheEntry<Object, Object>> previousBatch = batchRef.getAndSet(batch);
+                   if (previousBatch.isEmpty())
+                      return Completable.complete();
+
+                   return completionStageToCompletable().apply(sendEntries(previousBatch, false));
+                }, 1)
+                .subscribe(new CompletableObserver() {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: I would say we could use http://reactivex.io/RxJava/javadoc/io/reactivex/Completable.html#subscribe-io.reactivex.functions.Action-io.reactivex.functions.Consumer- instead now, but I won't let that block integration.

**Comment by wburns** [comment]
Integrated into master, thanks @danberindei !

File: core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java (hunk 21 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java]: I was actually using that at first, but I switched to the anonymous class to cut back on the lambdas :)

TBH maintaining the subscription was easier that way, because that overload is returning a `Disposable`, but I didn't want to mess with subscriptions any more.
